### ISSUE TITLE
webassembly for something else than browser

### ISSUE THREAD

**Issue by adridadou** [scala-native/scala-native#1206]
This is a question to understand if WebAssembly is out of the question or just out of scope.

Here is some context for this:
I am working on a blockchain project. My idea is to be able to compile my library to the JVM, JS and to the Ethereum Virtual Machine. Right now, the EVM uses a custom set of op codes but they are going to move to a wasm based virtual machine, called eWasm
https://github.com/ewasm/design

I think that one of the power of Scala is the fact that you can compile it to multiple bytecodes. If we were able to compile it to wasm as well, that would be very helpful for me and other projects too. 

So is it possible to do that are are there technical issues that are impossible to overcome?
If it is, how much work do you guys think we are talking about? I'd love to help as much as I can or try and find someone who can but I first want to understand how much work we are talking about here.

**Comment by adridadou** [comment]
p.s. scala.js also won't support web assembly. In some issue in their github they say that it is very unlikely that they will handle it

**Comment by mkotsbak** [comment]
See #603.

**Comment by adridadou** [comment]
I saw this already. 
It is written there that the reason why scala.native doesn't want to handle webassembly is that if you want to run scala code in the browser, you already have scala.js.

What I say is, Ethereum (and I guess others in the future) will base their virtual machine on WebAssembly and having scala native handle WebAssembly will make it compatible with Ethereum as smart contract.

**Comment by densh** [comment]
I believe that this is a legitimate feature request. It doesn't have any overlap with Scala.js and serves a legitimate new use case. Are there are any other non-web uses case of webassembly in the wild yet?

**Comment by rwhaling** [comment]
Francesc Campoy had this great example of running WebAssembly programs inside of a BigQuery UDF: 
https://blog.sourced.tech/post/calling-c-functions-from-bigquery/

**Comment by adridadou** [comment]
I was wondering, I saw this issue being closed https://github.com/WebAssembly/binaryen/issues/1447
does that mean that if we have an LLVM representation of our program, we can compile it to WebAssembly? Isn't that already the case with scala-native?

**Comment by ekrich** [comment]
We may need to update the FAQ if we plan to support Web Assembly for some use cases. http://www.scala-native.org/en/latest/faq.html#faq

**Comment by mkotsbak** [comment]
Of course this could be useful. Depending on the use case, the performance could be better by running as wasm than Scala.js. They can also be combined with a wasm module running high performance code as a webworker while GUI code is written in Scala.js.

Rust is already supported with wasm: https://rustwasm.github.io/book/ (although it does not depend on a GC).

**Comment by banshee** [comment]
I think for a bunch of use cases the never-collect garbage collector is completely viable.  For example, EOS doesn't even free memory when you're using C++ (free is essentially a no op, although destructors do run).

**Comment by scalavision** [comment]
I think WASM will become ``the way`` to deploy serverless solutions:

* https://blog.cloudflare.com/webassembly-on-cloudflare-workers/

Here is an example written in Rust:

* https://new.blog.cloudflare.com/cloudflare-workers-as-a-serverless-rust-platform/

To serve functions like this, I wonder if we need garbage collection at all? It's supposed to be stateless, fire and forget isn't it ... ?

If this comes true:

* https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-overview.md

There would exist a sandboxed system for deploying apps everywhere on any personal / home, premise or cloud platform.

I think however, that it is much more important to complete the really amazing work that is happening at the moment :-)

**Comment by bblfish** [comment]
From afar it looks like WebAssembly is really a form of Machine Code. It runs in the browser but can also run in the server with node.js.  
It looks a lot closer to what Scala Native is doing than what Scala-JS is doing. In a way it looks like it is very close to a new version of Java Bytecode, so one could argue Scala pure should compile to Web Assembly too (see a recent intro to [Web Assembly from June this year](https://www.youtube.com/watch?v=qDTpLKJ6e4M). 
But I suppose Java does not just come with bytecode it comes with whole libraries too.

**Comment by mkotsbak** [comment]
@bblfish if you need Java libraries, it is probably better to wait for solutions using GraalVM: https://github.com/oracle/graal/issues/1358

**Comment by bblfish** [comment]
@mkotsbak I don't think I mentioned needing Java libraries. :-) 

I am very happy with Scala, where it targets JVM bytecode. I am also very happy with ScalaJS as a better JS.
The above does leave an opening for Scala to compete with Rust by also producing Web Assembly code. If Scala-Native could compete at the Web Assembly level, it would make an amazingly strong argument for programming in Scala all the way around. (especially if Scala-native at least takes on linear types)

**Comment by mkotsbak** [comment]
@bblfish yes, of course using only Scala libraries as required in Scala.js and Native would be the best, but unfortunately there is a lot of missing features as much of the standard library people use is Java or Scala libraries that depend on Java. To compete with Rust in the short term I think Java bytecode to wasm would be the best.

**Comment by adridadou** [comment]
I just want to bring back my argument here :) A really good reason to have WebAssembly for scala is that Ethereum is developing their EWasm (Ethereum WebAssembly) VM for their next big release and it would be amazing to be able to deploy scala code there. Rust can do it already and has an edge but I'm personally looking for a way to have the exact same code for backend (jvm) embedded (native) browser (scala.js) and smart contract (WebAssembly)

**Comment by ekrich** [comment]
There was this little demo - https://github.com/shadaj/scala-native-wasm

Could this feature help? https://github.com/scala-native/scala-native/pull/2145

**Comment by bblfish** [comment]
Btw, here is an interesting paper on writing an RDF library in Rust, and having it compile to WASM, what needs to be done to have it be efficient to work together with JS. 

"WasmTree: Web Assembly for the Semantic Web"
https://openreview.net/pdf?id=CXLmXMb2TJ

The more processing that ends up happening in the WASM the bigger the efficiency gains - as one recovers the efficiency loss that happens in the communication layer. So placing a reasoning engine in WASM could make a huge difference for example.
